// nativescript/lib/common/services/lock-service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const lockfile = require("lockfile");
const path = require("path");
const decorators_1 = require("../decorators");
class LockService {
    constructor($fs, $settingsService, $processService) {
        this.$fs = $fs;
        this.$settingsService = $settingsService;
        this.$processService = $processService;
        this.currentlyLockedFiles = [];
        this.$processService.attachToProcessExitSignals(this, () => {
            const locksToRemove = _.clone(this.currentlyLockedFiles);
            _.each(locksToRemove, lock => {
                this.unlock(lock);
            });
        });
    }
    get defaultLockFilePath() {
        return this.getAbsoluteLockFilePath("lockfile.lock");
    }
    getAbsoluteLockFilePath(relativeLockFilePath) {
        return path.join(this.$settingsService.getProfileDir(), relativeLockFilePath);
    }
    get defaultLockParams() {
        const lockParams = {
            retryWait: 100,
            retries: 100,
            stale: 30 * 1000,
        };
        return lockParams;
    }
    executeActionWithLock(action, lockFilePath, lockOpts) {
        return __awaiter(this, void 0, void 0, function* () {
            const resolvedLockFilePath = yield this.lock(lockFilePath, lockOpts);
            try {
                const result = yield action();
                return result;
            }
            finally {
                this.unlock(resolvedLockFilePath);
            }
        });
    }
    lock(lockFilePath, lockOpts) {
        const { filePath, fileOpts } = this.getLockFileSettings(lockFilePath, lockOpts);
        this.currentlyLockedFiles.push(filePath);
        this.$fs.ensureDirectoryExists(path.dirname(filePath));
        return new Promise((resolve, reject) => {
            lockfile.lock(filePath, fileOpts, (err) => {
                err ? reject(new Error(`Timeout while waiting for lock "${filePath}"`)) : resolve(filePath);
            });
        });
    }
    unlock(lockFilePath) {
        const { filePath } = this.getLockFileSettings(lockFilePath);
        _.remove(this.currentlyLockedFiles, e => e === lockFilePath);
        lockfile.unlockSync(filePath);
    }
    getLockFileSettings(filePath, fileOpts) {
        if (filePath && !path.isAbsolute(filePath)) {
            filePath = this.getAbsoluteLockFilePath(filePath);
        }
        filePath = filePath || this.defaultLockFilePath;
        fileOpts = fileOpts || this.defaultLockParams;
        return {
            filePath,
            fileOpts
        };
    }
}
__decorate([
    decorators_1.cache()
], LockService.prototype, "defaultLockFilePath", null);
exports.LockService = LockService;
$injector.register("lockService", LockService);
$injector.register("lockfile", LockService);
