// nativescript/lib/commands/clean-app.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_base_1 = require("./command-base");
class CleanAppCommandBase extends command_base_1.ValidatePlatformCommandBase {
    constructor($options, $projectData, $platformService, $errors, $devicePlatformsConstants, $platformsData) {
        super($options, $platformsData, $platformService, $projectData);
        this.$errors = $errors;
        this.$devicePlatformsConstants = $devicePlatformsConstants;
        this.allowedParameters = [];
        this.$projectData.initializeProjectData();
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const appFilesUpdaterOptions = {
                bundle: !!this.$options.bundle,
                release: this.$options.release,
                useHotModuleReload: false
            };
            const platformInfo = {
                appFilesUpdaterOptions,
                platform: this.platform.toLowerCase(),
                config: this.$options,
                platformTemplate: this.$options.platformTemplate,
                projectData: this.$projectData,
                env: this.$options.env
            };
            return this.$platformService.cleanDestinationApp(platformInfo);
        });
    }
    canExecute(args) {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.$platformService.isPlatformSupportedForOS(this.platform, this.$projectData)) {
                this.$errors.fail(`Applications for platform ${this.platform} can not be built on this OS`);
            }
            const result = yield _super("canExecuteCommandBase").call(this, this.platform);
            return result;
        });
    }
}
exports.CleanAppCommandBase = CleanAppCommandBase;
class CleanAppIosCommand extends CleanAppCommandBase {
    constructor($options, $devicePlatformsConstants, $platformsData, $errors, $platformService, $projectData) {
        super($options, $projectData, $platformService, $errors, $devicePlatformsConstants, $platformsData);
        this.$options = $options;
        this.$devicePlatformsConstants = $devicePlatformsConstants;
        this.$platformsData = $platformsData;
        this.$errors = $errors;
    }
    get platform() {
        return this.$devicePlatformsConstants.iOS;
    }
}
exports.CleanAppIosCommand = CleanAppIosCommand;
$injector.registerCommand("clean-app|ios", CleanAppIosCommand);
class CleanAppAndroidCommand extends CleanAppCommandBase {
    constructor($options, $devicePlatformsConstants, $platformsData, $errors, $platformService, $projectData) {
        super($options, $projectData, $platformService, $errors, $devicePlatformsConstants, $platformsData);
        this.$options = $options;
        this.$devicePlatformsConstants = $devicePlatformsConstants;
        this.$platformsData = $platformsData;
        this.$errors = $errors;
    }
    get platform() {
        return this.$devicePlatformsConstants.Android;
    }
}
exports.CleanAppAndroidCommand = CleanAppAndroidCommand;
$injector.registerCommand("clean-app|android", CleanAppAndroidCommand);
